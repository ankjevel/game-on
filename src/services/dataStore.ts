type Check = <T>(result: MaybeNull<T>) => boolean

import uuid from 'uuid'
import * as userService from './user'
import { parse } from '../utils'
import redis from '../adapters/redis'

import {
  Action,
  GetResult,
  Group,
  isAction,
  isGroup,
  isUser,
  StoreTypes,
  Types,
  User,
} from '../types/dataStore'

export { Action, GetResult, Group, StoreTypes, Types, User, Check }

const client = redis()

export const newId = (type: StoreTypes) => `${type}:${uuid.v4()}`

export const stripTag = (input: string) => input.split(':')[1] || ''

export const checkDuplicate = async <T extends Types>(input: T) => {
  if (isAction(input) || isGroup(input)) {
    return
  }

  if (isUser(input)) {
    return userService.checkDuplicate(input)
  }
}

export const create = async <T extends Types>(
  type: StoreTypes,
  format: (id: string) => T
): Promise<T> => {
  const id = newId(type)

  if (await client.get(id)) {
    return await create(type, format)
  }

  const object = format(id)

  await checkDuplicate<T>(object)

  await client.set(
    id,
    typeof object === 'string' ? object : JSON.stringify(object)
  )

  return object
}

export const all = async (prefix: StoreTypes) =>
  await client.keys(`${prefix}:*`)

export const update = async <T extends Types>(
  id: T['id'],
  data: T,
  type: StoreTypes
) => {
  if (id !== data.id) {
    return
  }

  const prev = await getWrapper({ id, type })
  if (prev == null) {
    return
  }

  await client.set(
    id,
    JSON.stringify({
      ...prev,
      ...data,
    })
  )
}

export const get = async <T extends Types>({
  type,
  id,
  check,
}: {
  type: StoreTypes
  id: string
  check: Check
}): Promise<MaybeNull<T>> => {
  const query =
    id.includes(':') === false
      ? type == null
        ? null
        : `${type}:${id}`
      : type != null && id.split(':')[0] != type
      ? null
      : id

  if (query == null || !query.startsWith(`${type}:`)) {
    return null
  }

  const result = await client.get(query)
  if (result == null) {
    return null
  }

  try {
    const parsed = parse<T>(result)
    return check(parsed) ? parsed : null
  } catch (error) {
    console.error(result, error)
    return null
  }
}

export const getWrapper = async <T extends Types>({
  type,
  id,
}: {
  type: StoreTypes
  id: string
}): Promise<MaybeNull<T>> => {
  let check: Check

  switch (type) {
    case StoreTypes.Action:
      check = isAction
      break
    case StoreTypes.Group:
      check = isGroup
      break
    case StoreTypes.User:
      check = isUser
      break
    default:
      throw new Error('missing type')
  }

  return get<T>({ id, type, check })
}
