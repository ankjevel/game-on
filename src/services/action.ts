import { UserWithOutPassword, Action, Group, User } from 'dataStore'
import {
  StoreTypes as Type,
  ActionRunning,
  NewAction,
  checkId,
  isNewAction,
  NewActionEnum as NAE,
  NewActionEnum,
  UserSummary,
} from '../types/dataStore'
import { Message } from '../types/action'
import { getWrapper as getFromStore, update } from './dataStore'
import { pushSession } from './session'
import { parse, hasProp, clone, debug } from '../utils'
import mainLoop from './messageListener'

type QueryNext = {
  start: number
  userID: User['id']
  group: Group
  action: ActionRunning
  nextIndex: (current: number) => number
  check?: (value: NAE) => Boolean
}

const CHANNEL = 'message'

const isMessage = (input: Message | any): input is Message =>
  input != null &&
  (hasProp<any>(input, 'actionID') &&
    checkId(input.actionID, Type.ActionRunning)) &&
  (hasProp<any>(input, 'groupID') && checkId(input.groupID, Type.Group)) &&
  (hasProp<any>(input, 'userID') && checkId(input.userID, Type.User)) &&
  (hasProp<any>(input, 'newAction') && isNewAction(input.newAction))

const push = async (message: Message) => {
  if (!isMessage(message)) {
    return
  }

  await pushSession({ channel: CHANNEL, message: JSON.stringify(message) })
}

export const newAction = async ({
  actionID,
  groupID,
  userID,
  newAction,
}: {
  actionID: Action['id']
  groupID: Group['id']
  userID: UserWithOutPassword['id']
  newAction: NewAction
}) => {
  const action = await getFromStore<ActionRunning>({
    id: actionID,
    type: Type.ActionRunning,
  })

  const group = await getFromStore<Group>({
    id: groupID,
    type: Type.Group,
  })

  if (
    action == null ||
    group == null ||
    group.users.some(user => user.id === userID) === false ||
    group.action !== action.id ||
    (action.round === 4 &&
      group.owner !== userID &&
      (newAction.type !== NAE.Winner && newAction.type !== NAE.Draw))
  ) {
    return
  }

  await push({ actionID, userID, groupID, newAction })
}

const applyAction = (userAction: NAE, action: ActionRunning) =>
  action.round === 0 ? NAE.Bet : userAction

const getPlayer = ({
  start,
  group,
  action,
  userID,
  nextIndex,
  check,
}: QueryNext) => {
  const max = group.users.length
  let i = 0
  let run = true
  let nextUserIndex: MaybeNull<number> = start

  do {
    if (++i >= max) {
      console.log(action.id, 'cant find next player')
      run = false
      nextUserIndex = null
      break
    }

    nextUserIndex = nextIndex(nextUserIndex)

    const user = group.users[nextUserIndex]
    if (
      user == null ||
      user.id == null ||
      action.turn[user.id] == null ||
      userID === user.id
    ) {
      continue
    }

    const { status } = action.turn[user.id]
    if (
      check == null
        ? status !== NAE.Fold && status !== NAE.AllIn
        : check(status)
    ) {
      run = false
      break // found user
    }
  } while (run)

  return nextUserIndex
}

const raisePot = ({
  raise = 0,
  currentAnte,
  playerAnte,
  user,
  action,
  player,
}: {
  raise?: number
  currentAnte: number
  playerAnte: number
  user: Group['users'][0]
  action: ActionRunning
  player: UserSummary
}) => {
  const sum = currentAnte - playerAnte + raise
  if (user.sum <= sum) {
    console.log(action.id, 'cant update, not enough funds (maybe all-in?)')
    return false
  }

  user.sum -= sum
  player.bet += sum
  action.pot += sum

  return true
}

const handleUpdate = async (
  action: ActionRunning,
  group: Group,
  { newAction, userID }: Message
) => {
  const currentBig = clone(action.big)
  const isBig = userID === currentBig
  const currentAnte = action.turn[action.big].bet
  const userAction =
    newAction.type === NAE.None && action.queued[userID]
      ? action.queued[userID]
      : newAction
  const userIndex = group.users.findIndex(({ id }) => id === userID)
  const user = group.users[userIndex] as Group['users'][0]
  const player = action.turn[userID]
  const currentStatus = clone(player.status)
  const playerAnte = player.bet

  debug.action({ action, group, newAction, userID })

  if (!player) {
    console.log(action.id, `missing player ${userID}`)
    return
  }

  const nextUserIndex = getPlayer({
    start: userIndex,
    userID,
    group,
    action,
    nextIndex: current => (current + 1) % group.users.length,
  })

  switch (userAction.type) {
    case NAE.None: {
      console.log(action.id, 'no stored actions')
      return
    }

    case NAE.Bet: {
      if (player.status !== NAE.None) {
        console.log(action.id, 'bet is only used in start of round')
        return
      }

      if (
        !raisePot({
          currentAnte,
          playerAnte,
          user,
          action,
          player,
        })
      ) {
        return
      }

      player.status = NAE.Bet
      break
    }

    case NAE.Call: {
      if (
        !raisePot({
          currentAnte,
          playerAnte,
          user,
          action,
          player,
        })
      ) {
        return
      }

      player.status = applyAction(NAE.Call, action)
      break
    }

    case NAE.Raise: {
      if (userAction.value == null) {
        console.log(action.id, 'nothing raised')
        return
      }

      if (
        !raisePot({
          raise: userAction.value,
          currentAnte,
          playerAnte,
          user,
          action,
          player,
        })
      ) {
        return
      }

      action.big = userID
      player.status = applyAction(NAE.Raise, action)
      break
    }

    case NAE.Fold: {
      if (
        // edge case, where small blind folds in betting round
        action.round === 0 &&
        action.small === userID &&
        player.status === NAE.None
      ) {
        if (nextUserIndex == null || group.users[nextUserIndex] == null) {
          console.log(action.id, 'cant fold if no one else is in game :(')
          return
        }
        console.log(action.id, 'why you do this? folding')
        action.big = group.users[nextUserIndex].id
      }

      if (action.big === userID) {
        const nextUserIndex = getPlayer({
          start: userIndex,
          userID,
          group,
          action,
          nextIndex: current => (current + 1) % group.users.length,
          check: type => [NAE.Bet, NAE.Call, NAE.Check].includes(type),
        })

        console.log(action.id, 'handle edgecase where BIG folds')
        if (nextUserIndex == null || group.users[nextUserIndex] == null) {
          console.log(action.id, 'cant find NEXT better, maybe dont fold?')
          return
        } else {
          action.big = group.users[nextUserIndex].id
        }
      }

      player.status = NAE.Fold
      break
    }

    case NAE.Check: {
      if (!isBig) {
        const previousUserIndex = getPlayer({
          start: userIndex,
          userID,
          group,
          action,
          nextIndex: current => {
            --current

            if (current === -1) {
              current = group.users.length - 1
            }

            return current
          },
        })

        if (previousUserIndex == null) {
          console.log(action.id, 'cant find previous user')
          return
        }

        const prev = action.turn[group.users[previousUserIndex].id]

        if (prev.bet !== player.bet) {
          console.log(action.id, 'cant check, must raise (also not big)')
          return
        }
      }

      player.status = applyAction(NAE.Check, action)
      break
    }

    case NAE.AllIn: {
      const bet = player.bet + user.sum

      player.bet = bet
      action.pot += user.sum
      player.status = NAE.AllIn
      user.sum = 0

      if (action.sidePot == null) {
        action.sidePot = []
      }

      action.sidePot.push({
        id: userID,
        sum: bet,
      })

      break
    }

    default:
      console.log(action.id, `unhandled event: ${userAction.type}`)
      return
  }

  if (nextUserIndex == null || group.users[nextUserIndex] == null) {
    console.log(action.id, 'missing next player')
    return
  }

  const nextUserID = group.users[nextUserIndex].id
  console.log(action.id, `new button: ${action.button} => ${nextUserID}`)
  action.button = nextUserID

  if (action.round === 0) {
    if (Object.values(action.turn).every(turn => turn.status !== NAE.None)) {
      action.round += 1
      action.button = action.big
      console.log(action.id, 'should end betting round')
    }
  } else if (
    userAction.type !== NAE.Raise &&
    (userAction.type !== NAE.AllIn && action.big !== userID) &&
    nextUserID === currentBig
  ) {
    ++action.round
    if (action.round === 4) {
      console.log(action.id, 'showdown!\n\n')
    } else {
      console.log(action.id, 'new round\n\n')
    }
  }

  if (action.queued[userID]) {
    delete action.queued[userID]
  }

  if (
    // again, small becomes the big in first round
    action.round === 0 &&
    action.small === userID &&
    currentStatus === NAE.None &&
    // edge case where small folds at beginning of game
    player.status !== NAE.Fold &&
    action.big !== userID
  ) {
    console.log(action.id, 'set BIG <- SMALL')
    action.big = userID
  }

  maybeEnd: if (
    Object.values(action.turn).filter(
      x => x.status !== NAE.Fold && x.status !== NAE.AllIn
    ).length <= 1
  ) {
    if (
      Object.values(action.turn).filter(x => x.status === NAE.AllIn).length >= 1
    ) {
      action.round = 4
      console.log(action.id, 'all folded; contains all-in')
      break maybeEnd
    }

    console.log(action.id, `all folded; winner ${action.big}`)
    await handleEndRound(action, group, {
      type: NAE.Winner,
      winners: [action.big],
    })
    debug.endAction({ action, group })
    return
  }

  debug.endAction({ action, group })

  await update(action.id, action, Type.ActionRunning)
  await update(group.id, group, Type.Group)
}

type ActionGroup = {
  action: ActionRunning
  group: Group
}

const findNext = (group: Group, startIndex: number, sum: number, tries = 0) => {
  const max = group.users.length

  if (tries > max) {
    return -1
  }

  const next = (startIndex + 1) % max

  if (group.users[next].sum < sum) {
    console.log(`cant use user ${next}`)
    return findNext(group, next, sum, ++tries)
  }

  return next
}

const resetAction = ({
  action,
  group,
  pot = 0,
}: ActionGroup & {
  pot?: ActionRunning['pot']
}) => {
  const indexOfOldSmall = group.users.findIndex(
    user => user.id === action.small
  )

  const indexOfSmall = findNext(group, indexOfOldSmall, group.blind.small)
  const indexOfBig = findNext(group, indexOfSmall, group.blind.big)

  if (indexOfBig === -1 || indexOfSmall === -1) {
    console.log(
      action.id,
      `cant find next big|small [${indexOfBig},${indexOfSmall}]`
    )
    return
  }

  const newSmall = group.users[indexOfSmall]
  const newBig = group.users[indexOfBig]

  const turn = {
    [newSmall.id]: { bet: group.blind.small, status: NAE.None },
    [newBig.id]: { bet: group.blind.big, status: NAE.None },
  }

  group.users
    .filter(
      ({ id, sum }) =>
        id !== newSmall.id && id !== newBig.id && sum > group.blind.big
    )
    .forEach(user => {
      turn[user.id] = {
        bet: 0,
        status: NAE.None,
      }
    })

  group.users[indexOfSmall].sum -= group.blind.small
  group.users[indexOfBig].sum -= group.blind.big

  group.users = group.users.filter(user => user.sum > group.blind.big)

  action.pot = pot + group.blind.small + group.blind.big
  action.round = 0
  action.turn = turn
  action.sidePot = undefined
  action.button = newSmall.id
  action.queued = {}
  action.big = newBig.id
  action.small = newSmall.id
}

const handleEndRoundWithSidePot = async (
  action: ActionRunning,
  group: Group,
  newAction: Message['newAction']
) => {
  console.log({
    action,
    group,
    newAction,
  })
}

const handleEndRound = async (
  action: ActionRunning,
  group: Group,
  newAction: Message['newAction']
) => {
  if (action.sidePot != null) {
    return handleEndRoundWithSidePot(action, group, newAction)
  }

  let pot = 0
  switch (newAction.type) {
    case NAE.Draw: {
      const ids = Object.entries(action.turn)
        .filter(([, value]) => value.status !== NAE.Fold)
        .map(([id]) => id)

      const share = Math.floor(action.pot / ids.length)

      ids.forEach(id => {
        const user = group.users.find(user => user.id === id) || { sum: 0 }
        user.sum += share
      })

      pot = action.pot - share * ids.length
      break
    }

    case NAE.Winner: {
      const user = group.users.find(user => user.id === action.big) || {
        sum: 0,
      }
      user.sum += action.pot
      break
    }

    default:
      console.log(action.id, 'unhandled event', newAction)
      return
  }

  resetAction({ action, pot, group })

  await update(group.id, group, Type.Group)
  await update(action.id, action, Type.ActionRunning)
}

mainLoop(CHANNEL, async maybeMessage => {
  const message = parse<Message>(maybeMessage)
  if (!isMessage(message)) {
    console.info('not valid message, discarding', message)
    return
  }

  const group = await getFromStore<Group>({
    id: message.groupID,
    type: Type.Group,
  })

  const action = await getFromStore<ActionRunning>({
    id: message.actionID,
    type: Type.ActionRunning,
  })

  if (group == null) {
    console.info(`missing group: ${message.groupID}`)
    return
  }

  if (action == null) {
    console.info(`missing action: ${message.actionID}`)
    return
  }

  if (action.groupID !== group.id) {
    console.log(action.id, `wrong group: ${action.groupID} !== ${group.id}`)
    return
  }

  if (action.round === 4) {
    if (
      group.owner === message.userID &&
      (message.newAction.type === NewActionEnum.Draw ||
        message.newAction.type === NewActionEnum.Winner)
    ) {
      return handleEndRound(action, group, message.newAction)
    }

    console.log(action.id, 'group is in showdown')
    return
  }

  if (message.userID !== action.button) {
    action.queued[message.userID] = message.newAction

    console.log(
      action.id,
      `storing action ${message.newAction.type} for user ${message.userID}`
    )

    return await update(action.id, action, Type.ActionRunning)
  }

  return handleUpdate(action, group, message)
})
