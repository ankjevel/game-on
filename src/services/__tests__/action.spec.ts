import { ActionRunningWithSidePot, Message } from 'action'
import { UserWithOutPassword, Group, ActionRunning } from 'dataStore'

import * as dataStore from '../dataStore'
import * as drawFixture from '../__fixtures__/draw.fixture'
import * as betFoldAllInFixture from '../__fixtures__/bet-fold-all-in.fixture'

jest.mock('../dataStore')
jest.mock('../session')
jest.mock('../messageListener')

import * as actionService from '../action'
import { clone } from '../../utils'
import { checkHand } from '../cards'

let action: ActionRunningWithSidePot | ActionRunning
let users: UserWithOutPassword[]
let group: Group

let dataStoreMock: {
  [key: string]: jest.Mock
}

beforeEach(() => {
  dataStoreMock = dataStore as any
  console.log = jest.fn()
  console.info = jest.fn()
})

describe('#handleEndRoundWithSidePot', () => {
  beforeEach(() => {
    action = clone(drawFixture.action)
    users = clone(drawFixture.users)
    group = clone(drawFixture.group)
  })

  test.only('ALL should have the same amount', async () => {
    action.winners = [[users[2].id, users[0].id]]

    const resetAction = jest.spyOn(actionService, 'resetAction')

    expect(group.users.map(user => user.sum)).toEqual([140, 0, 0])

    resetAction.mockResolvedValue(true)

    await actionService.handleEndRoundWithSidePot(
      action as ActionRunningWithSidePot,
      group
    )
    expect(resetAction).toBeCalledTimes(1)
    expect(resetAction.mock.calls[0][0]).toMatchSnapshot()

    expect(group.users.map(user => user.sum)).toEqual([140, 140, 140])
  })
})

describe('#handleUpdate', () => {
  beforeEach(() => {
    action = clone(betFoldAllInFixture.action)
    users = clone(betFoldAllInFixture.users)
    group = clone(betFoldAllInFixture.group)
  })

  it('should not end round', async () => {
    const message = {
      newAction: {
        type: 'allIn',
      },
      userID: 'user:f7fefb97-1cd6-4507-b312-611d157a65bf',
    } as Message

    expect(action.round, 'starts at round 0').toEqual(0)

    await actionService.handleUpdate(action as ActionRunning, group, message)

    expect(action.round, 'still at round 0').toEqual(0)
    expect(dataStoreMock.update).toBeCalledTimes(2)

    expect(dataStoreMock.update.mock.calls).toEqual(
      expect.arrayContaining([
        expect.arrayContaining([action.id, action, 'action:running']),
        expect.arrayContaining([group.id, group, 'group']),
      ])
    )

    expect(action).toMatchSnapshot()
    expect(group).toMatchSnapshot()
  })
})

describe('#getPlayer', () => {})
