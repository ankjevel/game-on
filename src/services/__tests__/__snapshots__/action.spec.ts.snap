// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#handleEndRoundWithSidePot ALL should have the same amount 1`] = `
Object {
  "action": Object {
    "big": "user:63d94748-7134-4d5c-b3f9-612f9c0520c4",
    "button": "user:63d94748-7134-4d5c-b3f9-612f9c0520c4",
    "communityCards": Array [
      "ce",
      "b7",
      "ab",
      "d4",
      "c6",
    ],
    "deck": Array [
      null,
      "d2",
      "d3",
      null,
      "d5",
      "d6",
      "d7",
      "d8",
      "d9",
      "da",
      "db",
      "dd",
      null,
      "c1",
      "c2",
      "c3",
      "c4",
      "c5",
      null,
      "c7",
      "c8",
      "c9",
      "ca",
      "cb",
      "cd",
      null,
      null,
      null,
      "b3",
      "b4",
      "b5",
      "b6",
      null,
      "b8",
      null,
      "ba",
      "bb",
      "bd",
      "be",
      "a1",
      null,
      null,
      null,
      "a5",
      "a6",
      null,
      "a8",
      "a9",
      "aa",
      null,
      "ad",
      "ae",
    ],
    "groupID": "group:016650b1-8c11-41e3-b00f-a616469a3b1b",
    "id": "action:20ffb4b3-7fab-437c-be1e-cbd0efb8e00c",
    "pot": 0,
    "queued": Object {},
    "round": 4,
    "sidePot": Array [
      Object {
        "id": "user:63d94748-7134-4d5c-b3f9-612f9c0520c4",
        "sum": 140,
      },
      Object {
        "id": "user:f7fefb97-1cd6-4507-b312-611d157a65bf",
        "sum": 140,
      },
    ],
    "small": "user:63d94748-7134-4d5c-b3f9-612f9c0520c4",
    "turn": Object {
      "user:0f3bd090-e991-425a-8978-71ad27164510": Object {
        "bet": 0,
        "cards": Array [
          "a7",
          "b2",
        ],
        "hand": 8,
        "handParsed": Object {
          "highCards": Array [
            14,
            11,
            7,
            6,
            4,
            2,
          ],
          "onHand": Array [
            8,
          ],
          "parsed": Object {
            "cards": Object {
              "clubs": Array [
                4,
              ],
              "diamonds": Array [
                6,
                14,
              ],
              "hearts": Array [
                2,
                7,
              ],
              "spades": Array [
                7,
                11,
              ],
            },
            "flush": Object {
              "clubs": false,
              "diamonds": false,
              "hearts": false,
              "spades": false,
            },
            "fourOfAKinds": Array [],
            "pairs": Array [
              "7",
            ],
            "same": Object {
              "1": 0,
              "10": 0,
              "11": 1,
              "13": 0,
              "14": 1,
              "2": 1,
              "3": 0,
              "4": 1,
              "5": 0,
              "6": 1,
              "7": 2,
              "8": 0,
              "9": 0,
            },
            "straightFlushes": Array [],
            "threeOfAKinds": Array [],
          },
        },
        "status": "fold",
      },
      "user:63d94748-7134-4d5c-b3f9-612f9c0520c4": Object {
        "bet": 140,
        "cards": Array [
          "b9",
          "a4",
        ],
        "hand": 8,
        "handParsed": Object {
          "highCards": Array [
            14,
            11,
            9,
            7,
            6,
            4,
          ],
          "onHand": Array [
            8,
          ],
          "parsed": Object {
            "cards": Object {
              "clubs": Array [
                4,
              ],
              "diamonds": Array [
                6,
                14,
              ],
              "hearts": Array [
                7,
                9,
              ],
              "spades": Array [
                4,
                11,
              ],
            },
            "flush": Object {
              "clubs": false,
              "diamonds": false,
              "hearts": false,
              "spades": false,
            },
            "fourOfAKinds": Array [],
            "pairs": Array [
              "4",
            ],
            "same": Object {
              "1": 0,
              "10": 0,
              "11": 1,
              "13": 0,
              "14": 1,
              "2": 0,
              "3": 0,
              "4": 2,
              "5": 0,
              "6": 1,
              "7": 1,
              "8": 0,
              "9": 1,
            },
            "straightFlushes": Array [],
            "threeOfAKinds": Array [],
          },
        },
        "status": "allIn",
      },
      "user:f7fefb97-1cd6-4507-b312-611d157a65bf": Object {
        "bet": 140,
        "cards": Array [
          "de",
          "a3",
        ],
        "hand": 8,
        "handParsed": Object {
          "highCards": Array [
            14,
            11,
            7,
            6,
            4,
            3,
          ],
          "onHand": Array [
            8,
          ],
          "parsed": Object {
            "cards": Object {
              "clubs": Array [
                4,
                14,
              ],
              "diamonds": Array [
                6,
                14,
              ],
              "hearts": Array [
                7,
              ],
              "spades": Array [
                3,
                11,
              ],
            },
            "flush": Object {
              "clubs": false,
              "diamonds": false,
              "hearts": false,
              "spades": false,
            },
            "fourOfAKinds": Array [],
            "pairs": Array [
              "14",
            ],
            "same": Object {
              "1": 0,
              "10": 0,
              "11": 1,
              "13": 0,
              "14": 2,
              "2": 0,
              "3": 1,
              "4": 1,
              "5": 0,
              "6": 1,
              "7": 1,
              "8": 0,
              "9": 0,
            },
            "straightFlushes": Array [],
            "threeOfAKinds": Array [],
          },
        },
        "status": "allIn",
      },
    },
    "winners": Array [
      Array [
        "user:63d94748-7134-4d5c-b3f9-612f9c0520c4",
        "user:f7fefb97-1cd6-4507-b312-611d157a65bf",
      ],
    ],
  },
  "group": Object {
    "action": "action:20ffb4b3-7fab-437c-be1e-cbd0efb8e00c",
    "blind": Object {
      "big": 2,
      "small": 1,
    },
    "id": "group:016650b1-8c11-41e3-b00f-a616469a3b1b",
    "name": "super-duper-normal",
    "owner": "user:0f3bd090-e991-425a-8978-71ad27164510",
    "pub": true,
    "startSum": 140,
    "users": Array [
      Object {
        "id": "user:0f3bd090-e991-425a-8978-71ad27164510",
        "sum": 140,
      },
      Object {
        "id": "user:63d94748-7134-4d5c-b3f9-612f9c0520c4",
        "sum": 140,
      },
      Object {
        "id": "user:f7fefb97-1cd6-4507-b312-611d157a65bf",
        "sum": 140,
      },
    ],
  },
  "pot": 0,
}
`;

exports[`#handleEndRoundWithSidePot will not be a draw, if api decides (pair in 7 is not as high as kings) 1`] = `
Object {
  "action": Object {
    "big": "user:63d94748-7134-4d5c-b3f9-612f9c0520c4",
    "button": "user:63d94748-7134-4d5c-b3f9-612f9c0520c4",
    "communityCards": Array [
      "ce",
      "b7",
      "ab",
      "d4",
      "c6",
    ],
    "deck": Array [
      null,
      "d2",
      "d3",
      null,
      "d5",
      "d6",
      "d7",
      "d8",
      "d9",
      "da",
      "db",
      "dd",
      null,
      "c1",
      "c2",
      "c3",
      "c4",
      "c5",
      null,
      "c7",
      "c8",
      "c9",
      "ca",
      "cb",
      "cd",
      null,
      null,
      null,
      "b3",
      "b4",
      "b5",
      "b6",
      null,
      "b8",
      null,
      "ba",
      "bb",
      "bd",
      "be",
      "a1",
      null,
      null,
      null,
      "a5",
      "a6",
      null,
      "a8",
      "a9",
      "aa",
      null,
      "ad",
      "ae",
    ],
    "groupID": "group:016650b1-8c11-41e3-b00f-a616469a3b1b",
    "id": "action:20ffb4b3-7fab-437c-be1e-cbd0efb8e00c",
    "pot": 0,
    "queued": Object {},
    "round": 4,
    "sidePot": Array [
      Object {
        "id": "user:63d94748-7134-4d5c-b3f9-612f9c0520c4",
        "sum": 140,
      },
      Object {
        "id": "user:f7fefb97-1cd6-4507-b312-611d157a65bf",
        "sum": 140,
      },
    ],
    "small": "user:63d94748-7134-4d5c-b3f9-612f9c0520c4",
    "turn": Object {
      "user:0f3bd090-e991-425a-8978-71ad27164510": Object {
        "bet": 0,
        "cards": Array [
          "a7",
          "b2",
        ],
        "hand": 8,
        "handParsed": Object {
          "highCards": Array [
            14,
            11,
            7,
            6,
            4,
            2,
          ],
          "onHand": Array [
            8,
          ],
          "parsed": Object {
            "cards": Object {
              "clubs": Array [
                4,
              ],
              "diamonds": Array [
                6,
                14,
              ],
              "hearts": Array [
                2,
                7,
              ],
              "spades": Array [
                7,
                11,
              ],
            },
            "flush": Object {
              "clubs": false,
              "diamonds": false,
              "hearts": false,
              "spades": false,
            },
            "fourOfAKinds": Array [],
            "pairs": Array [
              "7",
            ],
            "same": Object {
              "1": 0,
              "10": 0,
              "11": 1,
              "13": 0,
              "14": 1,
              "2": 1,
              "3": 0,
              "4": 1,
              "5": 0,
              "6": 1,
              "7": 2,
              "8": 0,
              "9": 0,
            },
            "straightFlushes": Array [],
            "threeOfAKinds": Array [],
          },
        },
        "status": "fold",
      },
      "user:63d94748-7134-4d5c-b3f9-612f9c0520c4": Object {
        "bet": 140,
        "cards": Array [
          "b9",
          "a4",
        ],
        "hand": 8,
        "handParsed": Object {
          "highCards": Array [
            14,
            11,
            9,
            7,
            6,
            4,
          ],
          "onHand": Array [
            8,
          ],
          "parsed": Object {
            "cards": Object {
              "clubs": Array [
                4,
              ],
              "diamonds": Array [
                6,
                14,
              ],
              "hearts": Array [
                7,
                9,
              ],
              "spades": Array [
                4,
                11,
              ],
            },
            "flush": Object {
              "clubs": false,
              "diamonds": false,
              "hearts": false,
              "spades": false,
            },
            "fourOfAKinds": Array [],
            "pairs": Array [
              "4",
            ],
            "same": Object {
              "1": 0,
              "10": 0,
              "11": 1,
              "13": 0,
              "14": 1,
              "2": 0,
              "3": 0,
              "4": 2,
              "5": 0,
              "6": 1,
              "7": 1,
              "8": 0,
              "9": 1,
            },
            "straightFlushes": Array [],
            "threeOfAKinds": Array [],
          },
        },
        "status": "allIn",
      },
      "user:f7fefb97-1cd6-4507-b312-611d157a65bf": Object {
        "bet": 140,
        "cards": Array [
          "de",
          "a3",
        ],
        "hand": 8,
        "handParsed": Object {
          "highCards": Array [
            14,
            11,
            7,
            6,
            4,
            3,
          ],
          "onHand": Array [
            8,
          ],
          "parsed": Object {
            "cards": Object {
              "clubs": Array [
                4,
                14,
              ],
              "diamonds": Array [
                6,
                14,
              ],
              "hearts": Array [
                7,
              ],
              "spades": Array [
                3,
                11,
              ],
            },
            "flush": Object {
              "clubs": false,
              "diamonds": false,
              "hearts": false,
              "spades": false,
            },
            "fourOfAKinds": Array [],
            "pairs": Array [
              "14",
            ],
            "same": Object {
              "1": 0,
              "10": 0,
              "11": 1,
              "13": 0,
              "14": 2,
              "2": 0,
              "3": 1,
              "4": 1,
              "5": 0,
              "6": 1,
              "7": 1,
              "8": 0,
              "9": 0,
            },
            "straightFlushes": Array [],
            "threeOfAKinds": Array [],
          },
        },
        "status": "allIn",
      },
    },
    "winners": Array [
      Array [
        "user:f7fefb97-1cd6-4507-b312-611d157a65bf",
      ],
      Array [
        "user:63d94748-7134-4d5c-b3f9-612f9c0520c4",
      ],
    ],
  },
  "group": Object {
    "action": "action:20ffb4b3-7fab-437c-be1e-cbd0efb8e00c",
    "blind": Object {
      "big": 2,
      "small": 1,
    },
    "id": "group:016650b1-8c11-41e3-b00f-a616469a3b1b",
    "name": "super-duper-normal",
    "owner": "user:0f3bd090-e991-425a-8978-71ad27164510",
    "pub": true,
    "startSum": 140,
    "users": Array [
      Object {
        "id": "user:0f3bd090-e991-425a-8978-71ad27164510",
        "sum": 140,
      },
      Object {
        "id": "user:63d94748-7134-4d5c-b3f9-612f9c0520c4",
        "sum": 0,
      },
      Object {
        "id": "user:f7fefb97-1cd6-4507-b312-611d157a65bf",
        "sum": 280,
      },
    ],
  },
  "pot": 0,
}
`;

exports[`#handleUpdate should not end round 1`] = `
Object {
  "big": "user:f7fefb97-1cd6-4507-b312-611d157a65bf",
  "button": "user:0f3bd090-e991-425a-8978-71ad27164510",
  "communityCards": Array [],
  "deck": Array [
    "d1",
    "d2",
    "d3",
    "d4",
    "d5",
    "d6",
    "d7",
    "d8",
    "d9",
    "da",
    "db",
    "dd",
    "de",
    "c1",
    "c2",
    "c3",
    "c4",
    "c5",
    "c6",
    "c7",
    "c8",
    "c9",
    "ca",
    "cb",
    "cd",
    "ce",
    "b1",
    "b2",
    "b3",
    "b4",
    "b5",
    "b6",
    "b7",
    "b8",
    "b9",
    "ba",
    "bb",
    "bd",
    "be",
    "a1",
    "a2",
    "a3",
    "a4",
    "a5",
    "a6",
    "a7",
    "a8",
    "a9",
    "aa",
    "ab",
    "ad",
    "ae",
  ],
  "groupID": "group:016650b1-8c11-41e3-b00f-a616469a3b1b",
  "id": "action:7cc3e3d1-3f38-4b04-9d23-9de6372bc142",
  "pot": 22,
  "queued": Object {},
  "round": 0,
  "sidePot": Array [
    Object {
      "id": "user:f7fefb97-1cd6-4507-b312-611d157a65bf",
      "sum": 14,
    },
  ],
  "small": "user:0f3bd090-e991-425a-8978-71ad27164510",
  "turn": Object {
    "user:0f3bd090-e991-425a-8978-71ad27164510": Object {
      "bet": 2,
      "status": "bet",
    },
    "user:63d94748-7134-4d5c-b3f9-612f9c0520c4": Object {
      "bet": 2,
      "status": "fold",
    },
    "user:f7fefb97-1cd6-4507-b312-611d157a65bf": Object {
      "bet": 14,
      "status": "allIn",
    },
  },
}
`;

exports[`#handleUpdate should not end round 2`] = `
Object {
  "action": "action:7cc3e3d1-3f38-4b04-9d23-9de6372bc142",
  "blind": Object {
    "big": 2,
    "small": 1,
  },
  "id": "group:016650b1-8c11-41e3-b00f-a616469a3b1b",
  "name": "super-duper-normal",
  "owner": "user:0f3bd090-e991-425a-8978-71ad27164510",
  "startSum": 140,
  "users": Array [
    Object {
      "id": "user:0f3bd090-e991-425a-8978-71ad27164510",
      "sum": 28,
    },
    Object {
      "id": "user:63d94748-7134-4d5c-b3f9-612f9c0520c4",
      "sum": 370,
    },
    Object {
      "id": "user:f7fefb97-1cd6-4507-b312-611d157a65bf",
      "sum": 0,
    },
  ],
}
`;
